# News Headlines Sentiment

Use the news api to pull the latest news articles for bitcoin and ethereum and create a DataFrame of sentiment scores for each coin. 

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [3]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
load_dotenv()
from newsapi import NewsApiClient
import nltk
from nltk.sentiment.vader import SentimentIntensityAnalyzer
nltk.download('vader_lexicon')
analyzer = SentimentIntensityAnalyzer()


%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /Users/draganbogatic/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [4]:
nltk.download('stopwords')
nltk.download('punkt')
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/draganbogatic/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     /Users/draganbogatic/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/draganbogatic/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [5]:
# Read your api key environment variable
api_key = os.getenv("NEWS_API_KEY")

In [6]:
# Create a newsapi client
newsapi = NewsApiClient(api_key=api_key)

In [7]:
# Fetch the Bitcoin news articles
btc_headlines = newsapi.get_everything(q="bitcoin", language="en", page_size=100,sort_by="relevancy")
btc_articles = btc_headlines["articles"]
btc_articles[0]

{'source': {'id': 'wired', 'name': 'Wired'},
 'author': 'Timothy B. Lee, Ars Technica',
 'title': 'An Engineer Gets 9 Years for Stealing $10M From Microsoft',
 'description': 'The defendant tried—and failed—to use bitcoin to cover his tracks.',
 'url': 'https://www.wired.com/story/an-engineer-gets-9-years-for-stealing-dollar10m-from-microsoft/',
 'urlToImage': 'https://media.wired.com/photos/5fac6afb446b4639b3d5b8d8/191:100/w_1280,c_limit/Security-Microsoft-1229426260.jpg',
 'publishedAt': '2020-11-12T14:00:00Z',
 'content': "A former Microsoft software engineer from Ukraine has been sentenced to nine years in prison for stealing more than $10 million in store credit from Microsoft's online store. From 2016 to 2018, Volod… [+3307 chars]"}

In [8]:
# Fetch the Ethereum news articles
eth_headlines = newsapi.get_everything(q="ethereum", language="en", page_size=100,sort_by="relevancy")
eth_articles = eth_headlines["articles"]
eth_articles[0]

{'source': {'id': 'engadget', 'name': 'Engadget'},
 'author': 'Karissa Bell',
 'title': 'PayPal now lets all US users buy, sell and hold cryptocurrency',
 'description': 'PayPal is bringing its newly-announced support for cryptocurrency to all US accounts. It first announced plans to open cryptocurrency trading to US-based users in October, but until now it was only available to a small subset of PayPal account holders. That’s…',
 'url': 'https://www.engadget.com/paypal-opens-cryptocurrency-all-us-accounts-210541778.html',
 'urlToImage': 'https://o.aolcdn.com/images/dims?resize=1200%2C630&crop=1200%2C630%2C0%2C0&quality=95&image_uri=https%3A%2F%2Fs.yimg.com%2Fos%2Fcreatr-uploaded-images%2F2020-11%2F4e1af080-2528-11eb-bbdf-193e571d142a&client=amp-blogside-v2&signature=22ad23dde6ee3e667a8067a4c39e1962659cec92',
 'publishedAt': '2020-11-12T21:05:41Z',
 'content': 'PayPal is bringing its newly-announced support for cryptocurrency to all US accounts. It first announced plans to open cryptoc

In [9]:
# Create the Bitcoin sentiment scores DataFrame
btc_sentiments = []

for article in btc_articles:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        btc_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
        
    except AttributeError:
        pass
    
# Create DataFrame
btc_df = pd.DataFrame(btc_sentiments)

# Reorder DataFrame columns
cols = ["date", "text", "compound", "positive", "negative", "neutral"]
btc_df = btc_df[cols]

btc_df.head()

date                                               text  compound  \
0  2020-11-12  A former Microsoft software engineer from Ukra...   -0.6705   
1  2020-12-03  Visa has partnered with cryptocurrency startup...    0.6369   
2  2020-11-12  PayPal is bringing its newly-announced support...    0.2144   
3  2020-11-05  Two days ago, about $1 billion worth of bitcoi...    0.2263   
4  2020-11-20  In November 2017, after an absolutely massive,...    0.2023   

   positive  negative  neutral  
0     0.064     0.199    0.737  
1     0.162     0.000    0.838  
2     0.053     0.000    0.947  
3     0.049     0.000    0.951  
4     0.050     0.000    0.950

In [10]:
# Create the ethereum sentiment scores DataFrame
eth_sentiments = []

for article in eth_articles:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        eth_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
        
    except AttributeError:
        pass
    
# Create DataFrame
eth_df = pd.DataFrame(eth_sentiments)

# Reorder DataFrame columns
cols = ["date", "text", "compound", "positive", "negative", "neutral"]
eth_df = eth_df[cols]

eth_df.head()

date                                               text  compound  \
0  2020-11-12  PayPal is bringing its newly-announced support...    0.2144   
1  2020-11-23  FILE PHOTO: Representation of the Ethereum vir...    0.0000   
2  2020-11-23  FILE PHOTO: Representation of the Ethereum vir...    0.0000   
3  2020-11-23  LONDON (Reuters) - Digital currencies Ethereum...    0.4215   
4  2020-11-19  PayPal has launched the Generosity Network, a ...    0.8779   

   positive  negative  neutral  
0     0.053       0.0    0.947  
1     0.000       0.0    1.000  
2     0.000       0.0    1.000  
3     0.088       0.0    0.912  
4     0.318       0.0    0.682

In [11]:
# Describe the Bitcoin Sentiment
btc_df.describe()

compound   positive   negative    neutral
count  98.000000  98.000000  98.000000  98.000000
mean    0.144343   0.054541   0.019806   0.925643
std     0.330034   0.059863   0.042576   0.072958
min    -0.670500   0.000000   0.000000   0.716000
25%     0.000000   0.000000   0.000000   0.861000
50%     0.000000   0.050000   0.000000   0.945500
75%     0.421500   0.108500   0.000000   1.000000
max     0.811800   0.198000   0.215000   1.000000

In [12]:
# Describe the Ethereum Sentiment
eth_df.describe()

compound   positive  negative    neutral
count  97.000000  97.000000  97.00000  97.000000
mean    0.231799   0.075557   0.02100   0.903454
std     0.372525   0.079041   0.04529   0.088017
min    -0.670500   0.000000   0.00000   0.653000
25%     0.000000   0.000000   0.00000   0.848000
50%     0.226300   0.074000   0.00000   0.912000
75%     0.510600   0.132000   0.00000   1.000000
max     0.883400   0.347000   0.19600   1.000000

### Questions:

Q: Which coin had the highest mean positive score?

A: ETH has a mean positive score of 0.077646 compared to BTC mean positive score of 0.057571

Q: Which coin had the highest compound score?

A: Both BTC and ETH have the same Max compound score of 0.946800

Q. Which coin had the highest positive score?

A: ETH has the highest positive score of 0.347000 compared t0 BTC 0.297000

---

# Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word
2. Remove Punctuation
3. Remove Stopwords

In [13]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import regex
import re

In [14]:
text = btc_df['text'].astype('string')
text

0     A former Microsoft software engineer from Ukra...
1     Visa has partnered with cryptocurrency startup...
2     PayPal is bringing its newly-announced support...
3     Two days ago, about $1 billion worth of bitcoi...
4     In November 2017, after an absolutely massive,...
                            ...                        
93    Posted 
Bitcoin soared to a record high again...
94    By Reuters Staff
(Reuters) - - 1/ STICKY TALK...
95    Is it a currency? Is it an asset? Or is it per...
96    * New Zealand dollar set to reach more than tw...
97    (Fixes typo in the euro price in the 11th para...
Name: text, Length: 98, dtype: string

In [15]:
# Expand the default stopwords list if necessary


In [16]:
# Complete the tokenizer function
def tokenizer(text):

    """Tokenizes text."""

    # Create a list of the words
    sentences = [sent_tokenize(article) for article in text]
    words = [word_tokenize(article) for article in text]
    #print(words)

    # Remove punctuation using regex
    regex = re.compile("[^a-zA-Z ]")
    re_clean = regex.sub('', text)
    words_clean = word_tokenize(re_clean)

    # Remove the stop words
    sw = set(stopwords.words('english'))
    first_pass = [word for word in words_clean if word not in sw]
   
    # Lemmatize Words into root words; convert the words to lowercase
    lemmatizer = WordNetLemmatizer()
    lem = [lemmatizer.lemmatize(word) for word in words_clean]
    tokens = [word.lower() for word in lem if word.lower() not in sw]
  

    return tokens
    

In [17]:
tokenizer(text[0])

['former',
 'microsoft',
 'software',
 'engineer',
 'ukraine',
 'ha',
 'sentenced',
 'nine',
 'year',
 'prison',
 'stealing',
 'million',
 'store',
 'credit',
 'microsofts',
 'online',
 'store',
 'volod',
 'char']

In [18]:
# Create a new tokens column for bitcoin
# YOUR CODE HERE

In [19]:
# Create a new tokens column for ethereum
# YOUR CODE HERE!

---

# NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [20]:
from collections import Counter
from nltk import ngrams

In [21]:
# Generate the Bitcoin N-grams where N=2
# YOUR CODE HERE!

In [22]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [23]:
# Use the token_count function to generate the top 10 words from each coin
def token_count(tokens, N=10):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [24]:
# Get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [25]:
# Get the top 10 words for Ethereum
# YOUR CODE HERE!

# Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [26]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [27]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [28]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

# Named Entity Recognition

In this section, you will build a named entity recognition model for both coins and visualize the tags using SpaCy.

In [29]:
import spacy
from spacy import displacy

In [30]:
# Optional - download a language model for SpaCy
# !python -m spacy download en_core_web_sm

In [31]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

## Bitcoin NER

In [32]:
# Concatenate all of the bitcoin text together
# YOUR CODE HERE!

In [33]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [34]:
# Render the visualization
# YOUR CODE HERE!

In [35]:
# List all Entities
# YOUR CODE HERE!

---

## Ethereum NER

In [36]:
# Concatenate all of the bitcoin text together
# YOUR CODE HERE!

In [37]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [38]:
# Render the visualization
# YOUR CODE HERE!

In [39]:
# List all Entities
# YOUR CODE HERE!